### PR TITLE
Enable recursive setting of the trainable parameter

### PR DESCRIPTION
**Pull request by Dapid**
Doing `model.trainable = False` doesn't do anything, since its layers seem to still be trainable. I propose the method `set_trainable` for both layers and models that will recursively set the parameter.

The main use case is when we have sub-models inside the model we want to freeze, since we won't have to manually loop through an unknown number of layer depths.

If you think this is a good addition, I'll change the examples, add some docs, and a test.

### CODE REVIEW

**Comment by Dref360** [comment]
Great addition, as you said need some docs and tests :)

**Comment by abnera** [comment]
Yay must definitely need this feature!!! Now if someone could also make use_learning_phase recursively so I can use TimeDistributed with BatchNormalization & Dropout I might finally get my model to work lol. Thanks @Dapid.

**Comment by Dref360** [comment]
@Dapid While you're into recursive stuff could you fix this in the same PR? Mostly copy this : https://github.com/Dapid/keras/blob/8fd785d4695aaea10cc4af30317ab7cdab8b5995/keras/engine/topology.py#L1925

In the Layers class

Otherwise, I'll do it in a separate PR no problem.

**Comment by t-ae** [comment]
I thought this feature is already implemented.
The code below doesn't update the weights.
Is there something wrong?

```python
#!/usr/bin/env python

import numpy as np
from keras.models import Sequential
from keras.layers import Dense, InputLayer

inner = Sequential([
    Dense(3, input_shape=(10,))
])

inner.trainable = False

model = Sequential([
    InputLayer((10,)),
    inner
])

model.compile("adam", "mse")

x = np.random.uniform(0, 1, [1000, 10])
y = np.random.uniform(0, 1, [1000, 3])

before = inner.get_weights().copy()
print(before)

model.fit(x, y, verbose=0)

after = inner.get_weights().copy()

print(after)
print(np.array_equal(before[0], after[0]))  # True
print(np.array_equal(before[1], after[1]))  # True
```

**Comment by Dapid** [comment]
@t-ae that is true for Dense layers, but not for all. BatchNormalization, for example, changes behaviour based on its `.trainable` property.

**Comment by abnera** [comment]
Let's get this merged :D. @Dapid can you do the same for use_learning_phase? It's also suffering from recursive chain issue. Where learning phase is not passed to submodels of submodels... 

Example TimeDistributed(Dropout()) got fix. but TimeDistributed(Sequential(Dropout)) or a cnn_model with dropout or BatchNormalization passed via TimeDistributed also will fail. 

Seems a lots of the bugs have to do with recursiveness and or chaining of models inside models.

**Comment by Dapid** [comment]
@abnera I'll track the use_learning_phase, but I think it should be a different PR to keep things neat. I'll ping you when done so you can review and help François.

**Comment by abnera** [comment]
@Dapid thanks. That should be no problem. I apologize for the trouble. I'm a researcher, not so good at the coding part, but good at finding bugs. I can help with testing and feedback. I'm also available on the slack channel as avn3r. 

PR: https://github.com/fchollet/keras/pull/6387 was supposed to address the use_learning_phase but still suffers from when model has submodels.

wrapper_test.py should test TimeDistirbuted(Sequential(Sequential(Dropout())) or something like that to ensure if can chain the parameters regardless of the number of submodels.

@fchollet: maybe in future version of Keras a redesing of how models pass meta-data between child and parents models might be needed? Just a thought, you the expert.

**Comment by fchollet** [comment]
The `trainable` attribute of a model behaves exactly like that of a layer (because models shared the same API as layers). It only affects users of the model. Not users of the underlying layers. And that's on purpose. You don't want a layer's state to change based on who uses it and how it's being used.

Consider:

```python
text_encoder = LSTM(...)

x = Input(...)
y = text_encoder(x)
y = Dense(...)(y)

text_clf_model = Model(x, y)

x2 = Input(...)
y2 = text_encoder(x2)

feature_extraction_model = Model(x, y)
feature_extraction_model.set_trainable(False)

# And now I can't train `text_encoder`? Really?
text_clf_model.fit(...)  # does not work
```

**Comment by abnera** [comment]
@fchollet Ok that may be its intended behavior. Then how can we best help each other as a community to allow a model to access its submodels and change the submodel layers trainable status.

Example:
```
base_model = VGG16(input_shape=(224, 224, 3))
top_model = Sequential([Dropout(.5, input_shape=base_model.output_shape[1:]), Dense(101)])
cnn_model = Model(inputs=base_model.input, outputs=top_model(base_model.output))
model = Sequential([TimeDistributed(cnn_model, input_shape=(3, 224, 224, 3))])
for layer in base_model.layers:
    layer.trainable = False
model.compile(optimizer=Nadam(lr=2e-3),
                      loss='categorical_crossentropy',
                      metrics=['accuracy'])
model.summary() # Only top model layers will be trained
model.save('model.h5')

model2 = load_model('model.h5') 
model2.summary() # No way to train base model layers now than we finished training top layers.
# Cannot re-create models and pass weights either will get a rank error based on how model was saved.
```

Now that we have trained top_model (submodel) of model. How can we load and train base_model (submodel) of model. I have also tried to create the model again and pass the weights. `model.load_weights(path)` but that won't work either since it leads to rank errors.

**Comment by fchollet** [comment]
> No way to train base model layers now than we finished training top layers.

Why? I don't understand your example.

**Comment by abnera** [comment]
@fchollet I apologize for lack of clarity. I believe @Dapid created this PR to set trainable parameters of submodels once when loading the model for a second time.

I'll use the example you provided in the Tensorflow Submit about Video Q&A but in here just video.
For video classification, you have recommended the usage of TimeDistributed. 
I first create my cnn_model (image model). The cnn_model is composed of two parts. The VGG16 base_model (without include top) and the top_model (classification layers).
In summary:
### Train Top Model Layers
1) Create CNN Model:
cnn_model = base_model (VGG16) + top_model(Dense layer)
2) Now I need to pass the cnn_model via the TimeDistributed.
TimeDistibuted(cnn_model)
3) freeze VGG16 layers since we will use imagenet wieghts while we train the new layers (top_model)
for layer in base_model.layers:
    layer.trainable = True
4) Compile & Train model: Save model with ModelCheckpoint callback.

### Fine-Tune base_model layers (VGG16)
1) model = load_model(path)
2) Cannot unfreeze the layers I already froze on the model. Since I do not have access to the sub-models anymore. So basically I cannot do fine-tuning transfer learning on this example.

I want to be able to unfreeze those VGG16 on the model I have already saved and trained. I do not want to convert to fix features because that last may lead to accuracy gains, so I want an end-to-end model training. But I have no way to unfreeze any layers of the submodels once I load the curent best model.

**Comment by fchollet** [comment]
> Cannot unfreeze the layers I already froze on the model

But, you can. Not sure about the specifics of your model, but if your combined model is a `Sequential` model made of two layers, a pre-trained CNN and a classifier on top, then you can do `model.layers[0] = False/True` to freeze/unfreeze the CNN. I don't understand your issue.

**Comment by abnera** [comment]
Just run this simple example in a notebook. 

model2 (the loaded model) will have 101k trainable parameters. 
if u do model2.trainable = True or model2.layer[0].trainable = True it will not make all weights trainable. It will leave it at 101k. That is the bug. Unless you have another way of making parameters trainable that I do not now. The question is simple given this example, try to print a model2.summary() and show all parameters are trainable. Mine still shows they aren't regardless of how many times I call model2.trainable = True.

```
base_model = VGG16(input_shape=(224, 224, 3))
top_model = Sequential([Dropout(.5, input_shape=base_model.output_shape[1:]), Dense(101)])
cnn_model = Model(inputs=base_model.input, outputs=top_model(base_model.output))
model = Sequential([TimeDistributed(cnn_model, input_shape=(3, 224, 224, 3))])
for layer in base_model.layers:
    layer.trainable = False
model.compile(optimizer=Nadam(lr=2e-3),
                      loss='categorical_crossentropy',
                      metrics=['accuracy'])
model.summary() # Only top model layers will be trained
model.save('model.h5')

model2 = load_model('model.h5') 
model2.summary() # No way to train base model layers now than we finished training top layers.
# Cannot re-create models and pass weights either will get a rank error based on how model was saved.
```

**Comment by fchollet** [comment]
So, `model2` is a `Sequential` model with a single layer, a `TimeDistributed` layer. Which wraps `cnn_model`. Which is just VGG16 with an added `Sequential` on top.

To retrieve all your guys:

```python
model2 = load_model(...)
time_distributed = model2.layers[0] 
cnn_model = time_distributed.layer
top_model = cnn_model.layers[-1]
all_other_vgg_layers = cnn_models.layers[:-1]
```

You can then unfreeze `all_other_vgg_layers` as you like. 

If you keep track of *what the code you have written yourself does*, there is nothing complicated about it. Btw there are cleaner ways to implement fine-tuning... take a look at the applications page in the docs.

**Comment by fchollet** [comment]
As for this PR: this is a function that does

```python
def set_trainable(layer, value):
   layer.trainable = value
   # case: container
   if hasattr(layer, 'layers'):
     for l in layer.layers:
       set_trainable(l, value)
   # case: wrapper (which is a case not covered by the PR)
   if hasattr(layer, 'layer'):
      layer.layer = value
```

What are the motivations for adding it as a method on all layers/containers? I don't get it.

**Comment by t-ae** [comment]
@Dapid 
I read `BatchNormalization` code, and couldn't find where `trainable` affects.
Didn't you mistake [`training`](https://github.com/fchollet/keras/blob/master/keras/layers/normalization.py#L188-L190) for `trainable`?

And we should consider `model.trainable = True` may break the untrainable layers inside.
I think it'll be better if setting only container's `trainable` works perfectly.

**Comment by abnera** [comment]
@fchollet Thanks, I guess I finally learn how to do this. I have been asking for ages. Maybe a great addition to docs.

As for this PR all I wanted was what I described above. Mainly it's not intuitive. From a practical standpoint, I would want a way for a parent model to force all it's children to trainable = True.

Example: `model2.trainable = True` should be sufficient for all weights to be unfreeze. Have a parent/children structure where is I make `sub_model.trainable = True` only the submodel weights get updated but if I want the parent (model.trainable = True); it should force it's children to trainable = True. That is the main feature I am personaly asking for. I thought that is what @Dapid aimed to do, but you would have to ask him.

**Comment by fchollet** [comment]
> I would want a way for a parent model to force all it's children to trainable = True.

That's reasonable. Maybe we can consider it. That would be a breaking API change, though.
